# Python Review — Birthday Problem

This notebook explains **Quiz 5: Birthday** from the Python review as if I were teaching it to another student.

## Problem

Assume that a year has 365 days and that every birthday is equally likely. In a class of 20 students, estimate the probability that **at least two students share a birthday** using a Monte Carlo simulation.

## Step 1: Import NumPy and set the random seed

The seed makes the random experiment reproducible. Using the same seed gives the same result each time.

In [1]:
import numpy as np
np.random.seed(2019)

## Step 2: Run 10,000 simulated classes

For each trial, I generate 20 random birthdays. I convert them to a `set`, which automatically removes duplicates. If the set has fewer than 20 values, at least two students share a birthday.

In [2]:
count = 0

for i in range(10000):
    birthdays = set(np.random.randint(1, 366, 20))
    l = len(birthdays)
    if l != 20:
        count += 1

probability = count / 10000
probability

0.4115

## Result

The simulation gives approximately **0.4115**, or **41.15%**. This means that a class of only 20 students already has about a 41% chance of containing a shared birthday.

## Why `set` is useful

Suppose the birthdays are `[10, 25, 25, 80]`. Converting this list to a set gives `{10, 25, 80}`. The length becomes 3 instead of 4, which tells us immediately that a duplicate existed.

This is simpler than comparing every birthday with every other birthday.

## What this teaches

A Monte Carlo simulation estimates a probability by repeating a random experiment many times. The Birthday Problem is a good example because a short simulation can answer a probability question without starting from a complicated formula.